browser_patches/firefox-beta/juggler/content/PageAgent.js

```diff
@@ -786,6 +787,37 @@ class PageAgent {
     frame.textInputProcessor().commitCompositionWith(text);
   }
 
+  async _setComposition({
+    text,
+    selectionStart,
+    selectionEnd,
+    replacementStart,
+    replacementEnd,
+  }) {
+    const frame = this._frameTree.mainFrame();
+    const tip = frame.textInputProcessor();
+    if (replacementStart !== undefined && replacementEnd !== undefined) {
+      const window = frame.domWindow();
+      const document = window.document;
+      const activeElement = document.activeElement;
+      if (activeElement !== null) {
+        if (
+          activeElement.nodeName === "TEXTAREA" ||
+          activeElement.nodeName === "INPUT"
+        ) {
+          activeElement.setSelectionRange(replacementStart, replacementEnd);
+        } else {
+          const selection = window.getSelection();
+          const anchorNode = selection.anchorNode;
+          selection.setBaseAndExtent(anchorNode, replacementStart, anchorNode, replacementEnd);
```

Comment:
This doesn't look quite right. Aside from the shadow dom issues, the macOS api this is modeled after expects this range to be in terms of the current root editable node. However your code is in terms of the leaf editable node.

You can see this code in TextInputHandler.mm : IMEInputHandler::SetMarkedText, which ends up in the web process as ContentEventHandler::OnSelectionEvent. I am working on a patch to fix this but I'm not sure if I should duplicate the logic in JavaScript or add a hack for us to call the c++ code directly.